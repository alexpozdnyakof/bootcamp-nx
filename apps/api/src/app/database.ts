import { Database } from 'sqlite3'

const dbFileName =
	process.env.NODE_ENV === 'test' ? `database-test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})

function migrate() {
	console.log({ __dirname })

	database.exec(
		'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
	)

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		addMigrationData()
	}
}

function addMigrationData() {
	console.log({ __dirname })
	database.exec(
		`INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
	)
}

export { database, migrate, addMigrationData }

// 	'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES (1, 'ホームページのリニューアル', '新しいフレッシュホームページの制作過程はこちら'), (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
