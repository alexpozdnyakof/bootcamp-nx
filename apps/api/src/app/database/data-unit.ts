type UniqueId = number

type UpdatedCreatedTime = {
	created: number
	updated: number
}

interface EntityPropsMap {
	task: {
		text: string
		done: boolean
	}
	task_list: {
		title: string
	}
	project: {
		title: string
	}
}

type EntityType = 'task' | 'task_list' | 'project'

type DataUnit<T extends EntityType> = {
	id: UniqueId
	parentId: UniqueId
	type: T
	properties: EntityPropsMap[T] & UpdatedCreatedTime
	children: Array<UniqueId>
}

type EntityUnit = DataUnit<'task'> | DataUnit<'task_list'> | DataUnit<'project'>

type AbstractStore = Array<EntityUnit>

type DataUnitDTO = Pick<
	DataUnit<keyof EntityPropsMap>,
	'parentId' | 'children' | 'type'
> & {
	properties: Omit<
		DataUnit<keyof EntityPropsMap>['properties'],
		'created' | 'updated'
	>
}

type DataUnitUpdateValue = Partial<
	Pick<DataUnitDTO, 'properties' | 'children' | 'parentId'>
> &
	Pick<EntityUnit, 'id'>

export {
	AbstractStore,
	EntityType,
	UniqueId,
	EntityUnit,
	DataUnitDTO,
	DataUnit,
	DataUnitUpdateValue,
	UpdatedCreatedTime,
}
