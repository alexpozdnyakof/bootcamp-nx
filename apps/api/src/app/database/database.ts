import { Database } from 'sqlite3'

import * as fs from 'fs'
import * as path from 'path'

const dbFileName = process.env.NODE_ENV === 'test' ? `test` : `database`

const database = new Database(`${__dirname}/${dbFileName}.sqlite`, err => {
	if (err) throw new Error(err.message)
})
const sqlUrl =
	process.env.NODE_ENV === 'test'
		? path.join(__dirname, '../../sql')
		: path.join(__dirname, '/sql')

function migrate() {
	console.log({ __dirname })

	database.exec(fs.readFileSync(sqlUrl.concat('/projects.sql')).toString())

	if (process.env.NODE_ENV === 'test') {
		console.log('test')
		seeds()
	}
}

function seeds() {
	console.log({ __dirname })
	database.exec(
		fs.readFileSync(sqlUrl.concat('/projects-migration.sql')).toString()
	)
}

export { database, migrate }

// 'CREATE TABLE IF NOT EXISTS projects (id INTEGER PRIMARY KEY AUTOINCREMENT,title VARCHAR(200) NOT NULL,description TEXT NULL)'
// `INSERT OR REPLACE INTO projects VALUES
// (1, 'ホームページのリニューアル',
// '新しいフレッシュホームページの制作過程はこちら'),
// (2, 'チェックアウトフォームのリファクタリング', '新しいチェックアウト フォームでユーザー エクスペリエンスを向上'), (3, '実験プロジェクト', null)`
