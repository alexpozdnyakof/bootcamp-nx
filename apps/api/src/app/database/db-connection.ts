import knex from 'knex'
import knexConfig from '../../knexfile'
const dbConnection = () => {
	const envConfig = knexConfig[process.env.NODE_ENV ?? 'development']
	return knex(envConfig)
}

export default dbConnection()
