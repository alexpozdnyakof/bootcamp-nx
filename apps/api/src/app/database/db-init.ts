import { Database } from 'sqlite3'
import { readFileSync, readdirSync } from 'node:fs'
import { join } from 'path'

function readSqlDirectory(subdir: string): Array<string> {
	const path = join(
		__dirname,
		process.env.NODE_ENV === 'test' ? '../../sql' : '/sql',
		subdir
	)
	return readdirSync(path).map(fileName =>
		readFileSync(join(path, '/', fileName)).toString()
	)
}

function migrate(dbInstance: Database) {
	const migrations = readSqlDirectory('/migrations')
	migrations.map(sqlString => dbInstance.exec(sqlString))
}

function seed(dbInstance: Database) {
	const seeds = readSqlDirectory('/seeds')
	seeds.map(sqlString => dbInstance.exec(sqlString))
}

export function InitSqlite() {
	const isTest = process.env.NODE_ENV === 'test'
	const dbpath = isTest ? ':memory:' : join(__dirname, '/database.sqlite')

	const database = new Database(dbpath, err => {
		if (err) throw new Error(err.message)
	})

	migrate(database)
	if (process.env.NODE_ENV === 'test') {
		seed(database)
	}

	return database
}

export default InitSqlite()
