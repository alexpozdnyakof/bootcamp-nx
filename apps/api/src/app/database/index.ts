export * from './data-unit'
export { default as database } from './db-connection'
