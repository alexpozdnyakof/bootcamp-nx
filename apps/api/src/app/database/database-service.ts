import database from './database'

type DbResult<T> = {
	status: 'success' | 'failed'
	result: T
}

export function DatabaseService() {
	return Object.freeze({
		async All<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.all(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Get<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.get(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		async Run<T>(
			query: string,
			params?: Array<string | number | boolean>
		): Promise<T> {
			return new Promise((resolve, reject) => {
				database.run(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
	})
}

export default DatabaseService()
