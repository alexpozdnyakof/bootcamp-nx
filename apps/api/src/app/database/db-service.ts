import { Database } from 'sqlite3'
import { InitSqlite } from './db-init'

type Params = Array<string | number | boolean>
interface DatabaseService {
	All<T>(query: string, params?: Params): Promise<Array<T>>
	Get<T>(query: string, params: Params): Promise<T>
	Run<T = undefined>(query: string, params?: Params): Promise<T>
}

export function DatabaseService(Database: Database): DatabaseService {
	function handleNotFound<T>(result: T | undefined): Promise<T> {
		return new Promise((resolve, reject) => {
			if (typeof result == 'undefined') {
				reject('Not Found')
			}
			resolve(result)
		})
	}

	async function Get<T>(query: string, params: Params): Promise<T> {
		return new Promise((resolve, reject) => {
			Database.get(query, params, (err, result: T) => {
				if (err) reject(err)
				resolve(result)
			})
		}).then(
			result => handleNotFound<T>(result as T),
			reason => Promise.reject(reason)
		)
	}

	return Object.freeze({
		async All<T>(query: string, params?: Params): Promise<Array<T>> {
			return new Promise((resolve, reject) => {
				Database.all(query, params ?? [], (err, result: Array<T>) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
		Get,
		async Run<T = undefined>(query: string, params?: Params): Promise<T> {
			return new Promise((resolve, reject) => {
				Database.run(query, params ?? [], (err, result: T) => {
					if (err) reject(err)
					resolve(result)
				})
			})
		},
	})
}

export default DatabaseService(InitSqlite())
