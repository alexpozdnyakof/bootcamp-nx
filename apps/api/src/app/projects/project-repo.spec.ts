import { database } from '../database'
import ProjectRepository from './project-repo'
describe('ProjectRepository', () => {
	beforeAll(async () => {
		await database.migrate.up({
			name: '20221129145930_create_project_table.js',
		})
		await database.seed.run({ specific: '01-project.js' })
	})
	it('should create new one project', async () => {
		const dto = {
			title: '新しい計画',
			description: '簡単な説明',
		}
		try {
			const pre = await ProjectRepository.GetAll(1)
			await ProjectRepository.Add(dto)
			const past = await ProjectRepository.GetAll(1)

			expect(past).toHaveLength(pre.length + 1)
			expect(past[past.length - 1]).toMatchSnapshot({
				created: expect.any(String),
				updated: expect.any(String),
				...dto,
				id: expect.any(Number),
			})
		} catch (e) {
			console.log(e)
		}
	})

	it('should update entity with id 1', async () => {
		const dto = {
			title: '新しい計画',
			description: '簡単な説明',
		}

		let { title, description } = await ProjectRepository.GetOne(1)

		expect({ title, description }).not.toEqual(dto)

		await ProjectRepository.Update(1, dto)
		;({ title, description } = await ProjectRepository.GetOne(1))

		expect({ title, description }).toEqual(dto)
	})
})
