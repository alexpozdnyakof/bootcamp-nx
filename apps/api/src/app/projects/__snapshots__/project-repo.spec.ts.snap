// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectRepository should create new one project 1`] = `
Object {
  "created": Any<String>,
  "description": "簡単な説明",
  "id": Any<Number>,
  "title": "新しい計画",
  "updated": Any<String>,
}
`;
